File: src/csharp-generator/Services/JsonParser.cs
using System.Text.Json;
using FluentValidation.Generator.Models;

namespace FluentValidation.Generator.Services;

/// <summary>
/// Service for parsing and validating JSON rule definition files
/// </summary>
public class JsonParser
{
    private readonly JsonSerializerOptions _jsonOptions;

    public JsonParser()
    {
        _jsonOptions = new JsonSerializerOptions
        {
            PropertyNameCaseInsensitive = true,
            ReadCommentHandling = JsonCommentHandling.Skip,
            AllowTrailingCommas = true,
        };
    }

    /// <summary>
    /// Reads and parses all JSON files from the specified directory
    /// </summary>
    /// <param name="inputPath">Path to directory containing JSON rule files</param>
    /// <returns>List of parsed validation definitions</returns>
    public async Task<List<ValidationDefinition>> ParseDirectoryAsync(string inputPath)
    {
        if (!Directory.Exists(inputPath))
        {
            throw new DirectoryNotFoundException($"Input directory not found: {inputPath}");
        }

        var jsonFiles = Directory.GetFiles(inputPath, "*.json", SearchOption.TopDirectoryOnly);

        if (jsonFiles.Length == 0)
        {
            Console.WriteLine($"Warning: No JSON files found in {inputPath}");
            return new List<ValidationDefinition>();
        }

        var definitions = new List<ValidationDefinition>();

        foreach (var filePath in jsonFiles)
        {
            try
            {
                var definition = await ParseFileAsync(filePath);
                definitions.Add(definition);
                Console.WriteLine($"✓ Parsed {Path.GetFileName(filePath)}");
            }
            catch (Exception ex)
            {
                Console.WriteLine($"✗ Error parsing {Path.GetFileName(filePath)}: {ex.Message}");
                throw;
            }
        }

        return definitions;
    }

    /// <summary>
    /// Parses a single JSON file into a ValidationDefinition
    /// </summary>
    /// <param name="filePath">Path to the JSON file</param>
    /// <returns>Parsed validation definition</returns>
    public async Task<ValidationDefinition> ParseFileAsync(string filePath)
    {
        if (!File.Exists(filePath))
        {
            throw new FileNotFoundException($"JSON file not found: {filePath}");
        }

        var jsonContent = await File.ReadAllTextAsync(filePath);

        var definition = JsonSerializer.Deserialize<ValidationDefinition>(
            jsonContent,
            _jsonOptions
        );

        if (definition == null)
        {
            throw new InvalidOperationException($"Failed to deserialize {filePath}");
        }

        ValidateDefinition(definition, filePath);

        return definition;
    }

    /// <summary>
    /// Validates that the parsed definition has all required fields
    /// </summary>
    private void ValidateDefinition(ValidationDefinition definition, string filePath)
    {
        var errors = new List<string>();

        if (string.IsNullOrWhiteSpace(definition.Entity))
        {
            errors.Add("Entity name is required");
        }

        if (string.IsNullOrWhiteSpace(definition.Namespace))
        {
            errors.Add("Namespace is required");
        }

        if (definition.Properties == null || definition.Properties.Count == 0)
        {
            errors.Add("At least one property must be defined");
        }
        else
        {
            for (int i = 0; i < definition.Properties.Count; i++)
            {
                var prop = definition.Properties[i];

                if (string.IsNullOrWhiteSpace(prop.Name))
                {
                    errors.Add($"Property[{i}]: Name is required");
                }

                if (string.IsNullOrWhiteSpace(prop.Type))
                {
                    errors.Add($"Property[{i}] ({prop.Name}): Type is required");
                }

                if (prop.Rules == null || prop.Rules.Count == 0)
                {
                    errors.Add(
                        $"Property[{i}] ({prop.Name}): At least one validation rule is required"
                    );
                }
                else
                {
                    for (int j = 0; j < prop.Rules.Count; j++)
                    {
                        var rule = prop.Rules[j];
                        if (string.IsNullOrWhiteSpace(rule.Validator))
                        {
                            errors.Add(
                                $"Property[{i}] ({prop.Name}), Rule[{j}]: Validator name is required"
                            );
                        }
                    }
                }
            }
        }

        if (errors.Any())
        {
            throw new InvalidOperationException(
                $"Validation errors in {Path.GetFileName(filePath)}:\n  - {string.Join("\n  - ", errors)}"
            );
        }
    }
}

File: src/typescript-generator/src/services/jsonParser.ts
import * as fs from 'fs/promises';
import * as path from 'path';
import { ValidationDefinition } from '../models/validationRuleModels';

/**
 * Service for parsing and validating JSON rule definition files
 */
export class JsonParser {
  /**
   * Reads and parses all JSON files from the specified directory
   */
  async parseDirectory(inputPath: string): Promise<ValidationDefinition[]> {
    try {
      const stats = await fs.stat(inputPath);
      if (!stats.isDirectory()) {
        throw new Error(`Input path is not a directory: ${inputPath}`);
      }
    } catch (error) {
      throw new Error(`Input directory not found: ${inputPath}`);
    }

    const files = await fs.readdir(inputPath);
    const jsonFiles = files.filter(file => file.endsWith('.json'));

    if (jsonFiles.length === 0) {
      console.log(`Warning: No JSON files found in ${inputPath}`);
      return [];
    }

    const definitions: ValidationDefinition[] = [];

    for (const fileName of jsonFiles) {
      try {
        const filePath = path.join(inputPath, fileName);
        const definition = await this.parseFile(filePath);
        definitions.push(definition);
        console.log(`✓ Parsed ${fileName}`);
      } catch (error) {
        console.error(`✗ Error parsing ${fileName}: ${(error as Error).message}`);
        throw error;
      }
    }

    return definitions;
  }

  /**
   * Parses a single JSON file into a ValidationDefinition
   */
  async parseFile(filePath: string): Promise<ValidationDefinition> {
    const fileContent = await fs.readFile(filePath, 'utf-8');
    const definition = JSON.parse(fileContent) as ValidationDefinition;

    this.validateDefinition(definition, filePath);

    return definition;
  }

  /**
   * Validates that the parsed definition has all required fields
   */
  private validateDefinition(definition: ValidationDefinition, filePath: string): void {
    const errors: string[] = [];

    if (!definition.entity || definition.entity.trim() === '') {
      errors.push('Entity name is required');
    }

    if (!definition.namespace || definition.namespace.trim() === '') {
      errors.push('Namespace is required');
    }

    if (!definition.properties || definition.properties.length === 0) {
      errors.push('At least one property must be defined');
    } else {
      definition.properties.forEach((prop, i) => {
        if (!prop.name || prop.name.trim() === '') {
          errors.push(`Property[${i}]: Name is required`);
        }

        if (!prop.type || prop.type.trim() === '') {
          errors.push(`Property[${i}] (${prop.name}): Type is required`);
        }

        if (!prop.rules || prop.rules.length === 0) {
          errors.push(`Property[${i}] (${prop.name}): At least one validation rule is required`);
        } else {
          prop.rules.forEach((rule, j) => {
            if (!rule.validator || rule.validator.trim() === '') {
              errors.push(`Property[${i}] (${prop.name}), Rule[${j}]: Validator name is required`);
            }
          });
        }
      });
    }

    if (errors.length > 0) {
      throw new Error(
        `Validation errors in ${path.basename(filePath)}:\n  - ${errors.join('\n  - ')}`
      );
    }
  }
}
